checks.d/mesos_master.py

```diff
@@ -0,0 +1,246 @@
+                self.SERVICE_CHECK_NEEDED = False
+            if status is AgentCheck.CRITICAL:
+                self.warning(msg)
+                return None
+
+        return r.json()
+
+    def _get_master_state(self, url, timeout):
+        return self._get_json(url + '/state.json', timeout)
+
+    def _get_master_stats(self, url, timeout):
+        if self.version >= Version('0.22.0'):
+            endpoint = '/metrics/snapshot'
+        else:
+            endpoint = '/stats.json'
+        return self._get_json(url + endpoint, timeout)
+
+    def _get_master_roles(self, url, timeout):
+        return self._get_json(url + '/roles.json', timeout)
+
+    def _check_leadership(self, url, timeout):
+        json = self._get_master_state(url, timeout)
+
+        if json is not None:
+            self.version = Version(json['version'])
+            if json['leader'] == json['pid']:
+                self.leader = True
+        else:
+            self.leader = False
+        return json
+
+    def check(self, instance):
+        if 'url' not in instance:
+            raise Exception('Mesos instance missing "url" value.')
+
+        url = instance['url']
+        instance_tags = instance.get('tags', [])
+        default_timeout = self.init_config.get('default_timeout', 5)
+        timeout = float(instance.get('timeout', default_timeout))
+
+        json = self._check_leadership(url, timeout)
+        if json:
+            tags = ['mesos_cluster:' + json['cluster'], 'mesos_pid:' + json['pid'], 'mesos_id:' + json['id'], 'mesos_node:master'] + instance_tags
+
+            if self.leader:
+                self.GAUGE('mesos.cluster.total_frameworks', len(json['frameworks']), tags=tags)
+
+                for framework in json['frameworks']:
+                    framework_tags = ['framework:' + framework['id'], 'framework_name:' + framework['name']] + tags
+                    self.GAUGE('mesos.framework.total_tasks', len(framework['tasks']), tags=framework_tags)
+                    resources = framework['used_resources']
+                    [v[1](self, v[0], resources[k], tags=framework_tags) for k, v in self.FRAMEWORK_METRICS.iteritems()]
+
+                json = self._get_master_roles(url, timeout)
+                if json is not None:
+                    for role in json['roles']:
+                        role_tags = ['mesos_role:' + role['name']] + tags
+                        self.GAUGE('mesos.role.frameworks', len(role['frameworks']), tags=role_tags)
+                        self.GAUGE('mesos.role.weight', role['weight'], tags=role_tags)
+                        [v[1](self, v[0], role['resources'][k], tags=role_tags) for k, v in self.ROLE_RESOURCES_METRICS.iteritems()]
+
+            json = self._get_master_stats(url, timeout)
+            if json is not None:
+                if self.leader:
+                    metrics = {}
+                    for d in (self.CLUSTER_TASKS_METRICS, self.CLUSTER_SLAVES_METRICS,
```

Comment:
Instead of merging all these dictionaries into one i would just iterate over the list of dictionaries, it would be more efficient and there shouldn't be any conflicts in the key names of these dictionaries anyway.